krm-functions/lib/kptfile/v1/kpt.go

```diff
@@ -0,0 +1,167 @@
+/*
+Copyright 2023 Nephio.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1
+
+import (
+	"errors"
+
+	"github.com/GoogleContainerTools/kpt-functions-sdk/go/fn"
+	kptv1 "github.com/GoogleContainerTools/kpt/pkg/api/kptfile/v1"
+	"sigs.k8s.io/yaml"
+)
+
+type KptFile interface {
+	// Unmarshal decodes the raw document within the in byte slice and assigns decoded values into the out value.
+	// it leverages the  "sigs.k8s.io/yaml" library
+	UnMarshal() (*kptv1.KptFile, error)
```

Comment:
My understanding is  this is mainly used  for CRUD operations for conditions in kpt files. if that's right my first thought looking at the interface is to separate the concerns into two separate interfaces . One for  file operations ( Marshall, Unmarshall etc) and one for CRUD on conditions.  It will keep the interface scope narrow and clear without any leak of abstractions. But there may be good reason why you may combined them into the same interface and that's fine as well. Just wanted to point out my thoughts.

Reply:
Yes what I saw as a pattern is that these methods are coming back all the time.

// Marshal serializes the value provided into a YAML document based on "sigs.k8s.io/yaml".
	// The structure of the generated document will reflect the structure of the value itself.
	Marshal() ([]byte, error)
	// ParseKubeObject returns a fn sdk KubeObject; if something failed an error
	// is returned
	ParseKubeObject() (*fn.KubeObject, error)

So I am going to move this into a general Parser interface which is generic across all of the libraries.

so we get this.

type KptFile interface {
        // generic parser interface which contains Marshal/ParseKubeObject
	parser.Parser
	// GetKptFile returns the Kptfile as a go struct
	GetKptFile() *kptv1.KptFile
	// SetConditions sets the conditions in the kptfile. It either updates the entry if it exists
	// or appends the entry if it does not exist.
	SetConditions(...kptv1.Condition)
	// DeleteCondition deletes the condition equal to the conditionType if it exists
	DeleteCondition(ct string)
	// GetCondition returns the condition for the given ConditionType if it exists,
	// otherwise returns nil
	GetCondition(ct string) *kptv1.Condition
	// GetConditions returns all the conditions in the kptfile. if not initialized it
	// returns an emoty slice
	GetConditions() []kptv1.Condition
}

Reply:
I created this PR. So if we approve this I can update this 
https://github.com/nephio-project/nephio/pull/84

this PR specifies the Parser interface